File: 69. Sqrt(x)/optimised.ts
function mySqrt(x: number): number {
	let [i, j] = [0, x];

	while (i <= j) {
		const mid = Math.floor((i + j) / 2);
		const square = mid * mid;

		if (square == x) {
			return mid;
		} else if (square < x) {
			i = mid + 1;
		} else {
			j = mid - 1;
		}
	}

	return j;
};

console.log(mySqrt(8));

File: 69. Sqrt(x)/optimised.py
#!/usr/bin/env python

def mySqrt(x: int) -> int:
    i, j = 0, x
    while i <= j:
        mid = (i + j) // 2
        square = mid * mid

        if square == x:
            return mid
        elif square < x:
            i = mid + 1
        else:
            j = mid - 1

    return j

print(mySqrt(8))
